### PR TITLE
Implement strstr() and strpbrk() from string.h

### PR DESCRIPTION
**Pull request by azymohliad**
This is my first attempt to contribute to this project and I'm not experienced in Rust, especially with its C FFI, so please review carefully. I'm not sure about some of my changes, I'll ask more specifically in diffs. Sorry for any mistakes in advance.

I have test failure on "math", but if I remove it from tests/Makefile, my test performs correctly.

P.S. If somebody is already doing it - please feel free to close my pull request. I'm just creating it for more convenient feedback.

### CODE REVIEW

File: src/string/src/lib.rs [hunk 1]
```diff
@@ -299,8 +299,22 @@ pub unsafe extern "C" fn strspn(s1: *const c_char, s2: *const c_char) -> c_ulong
 }
 
 #[no_mangle]
-pub extern "C" fn strstr(s1: *const c_char, s2: *const c_char) -> *mut c_char {
-    unimplemented!();
+pub unsafe extern "C" fn strstr(s1: *const c_char, s2: *const c_char) -> *mut c_char {
```
Comment (azymohliad) [inline src/string/src/lib.rs]: Although it seems logical to me, I'm not sure if I did the right thing when changed the signature to `unsafe`

File: src/string/src/lib.rs [hunk 2]
```diff
@@ -299,8 +299,22 @@ pub unsafe extern "C" fn strspn(s1: *const c_char, s2: *const c_char) -> c_ulong
...
+    let mut i = 0;
+    while *s1.offset(i) != 0 {
+        let mut j = 0;
+        while *s2.offset(j) != 0 && *s1.offset(j + i) != 0 {
+            if *s2.offset(j) != *s1.offset(j + i) {
+                break;
+            }
+            j += 1;
+            if *s2.offset(j) == 0 {
+                return s1.offset(i) as *mut c_char;
```
Comment (azymohliad) [inline src/string/src/lib.rs]: I'm not sure if casting `*const c_char` to `*mut c_char` is the right thing to do here

Comment (dlrobertson) [reply src/string/src/lib.rs]: Yup. Feels terrible, but it is what all the other libc's do.

File: src/string/src/lib.rs [hunk 3]
```diff
@@ -250,8 +250,15 @@ pub unsafe extern "C" fn strncpy(s1: *mut c_char, s2: *const c_char, n: usize) -
 }
 
 #[no_mangle]
-pub extern "C" fn strpbrk(s1: *const c_char, s2: *const c_char) -> *mut c_char {
-    unimplemented!();
+pub unsafe extern "C" fn strpbrk(s1: *const c_char, s2: *const c_char) -> *mut c_char {
+    let mut i = 0;
+    while *s1.offset(i) != 0 {
```
[the remark this answers is not in the document]
Comment (azymohliad) [reply src/string/src/lib.rs]: Thank you. I see we already have `strcspn()`. One thing: when characters from s2 isn't found in s1, `strpbrk()` should return null pointer, but `s1.offset(strcspn(s1, s2))` would be a valid pointer to `'\0'`. So I guess, it should be something like:
```
pub unsafe extern "C" fn strpbrk(s1: *const c_char, s2: *const c_char) -> *mut c_char {
    let mut p = s1.offset(strcspn(s1, s2)) as *mut c_char;
    if *p == 0 { p = ptr::null_mut; }
    ptr
}
```
Am I correct?

[the remark this answers is not in the document]
Comment (dlrobertson) [reply src/string/src/lib.rs]: Exactly. `musl` uses the following.

```c
*s ? (char*)s : 0
```

So the equivalent rust would be

```rust
if *p == 0 {
    ptr::nul_mut();
} else {
    p
}
```

[the remark this answers is not in the document]
Comment (azymohliad) [reply src/string/src/lib.rs]: Ok, I'll update it now. Thanks for the quick response

[the remark this answers is not in the document]
Comment (azymohliad) [reply src/string/src/lib.rs]: Seems `strcspn()` doesn't work properly yet. I got this error in tests:
```
src/string/src/lib.rs:133: attempt to shift left with overflow
make: *** [Makefile:48: run] Error 1
make: Leaving directory '/home/azymohliad/Projects/opensource/relibc/tests'
```
I'll check it more carefully after work

[the remark this answers is not in the document]
Comment (Tommoa) [reply src/string/src/lib.rs]: @azymohliad are you sure you're on the latest version of relibc? I committed a patch a while ago that should have fixed that.

If you have the patch, `strcspn()` should have something along the lines of:
```rust
    while *s2.offset(i) != 0 {
        byteset[(*s2.offset(i) as usize) / (8 * byteset.len())] |=
            1 << (*s2.offset(i) as usize % (8 * byteset.len()));
        i += 1;
    }
```
while the old version was:
```rust
     while *s2.offset(i) != 0 {
        byteset[(*s2.offset(i) as usize) / (8 * mem::size_of::<usize>())] |=
            1 << (*s2.offset(i) as usize % (8 * mem::size_of::<usize>()));
         i += 1;
     }
```

Testing `strcspn` and `strspn` hasn't worked so far for some reason, so its possible there's another error in there though (see #67)

[the remark this answers is not in the document]
Comment (Tommoa) [reply src/string/src/lib.rs]: It has occurred to me that the type of the byteset should be `[usize]`, not `[u8]`. The mod before the bitshift will only play nice 32 bit numbers. I'll send through a pull request.

[the remark this answers is not in the document]
Comment (azymohliad) [reply src/string/src/lib.rs]: @Tommoa great, thanks. With your fix and one more thing it works. This one more thing is: `strcspn()` has two loops and currently both are traversing `s2`. I think the second one should traverse `s1`. When I change it, my `strpbrk()` implementation using `strcspn()` works correctly.

[the remark this answers is not in the document]
Comment (azymohliad) [reply src/string/src/lib.rs]: @Tommoa would you prefer to add this fix to your pull request or should I add it here, or create a new issue for it?

[the remark this answers is not in the document]
Comment (Tommoa) [reply src/string/src/lib.rs]: Don't worry about it, I'll chuck it in mine. We really need to get #67 sorted out though, cause then this problem would have been fixed ages ago.

**Comment by Tommoa** [comment]
Oh yeah, about your failure on the math tests.

`openlibm` needs to be compiling with the `-fno-stack-protector`. I'd suggest going into the `openlibm` folder, running `make clean` and then going back out and running `./test.sh` again

**Comment by Tommoa** [comment]
I've figured out how to do some testing on `strcspn` and `strspn` and they're both outputting incorrect answers. I'm going to go figure out what's going on.

**Pull request merged**